guides/keras_cv/writing_custom_image_augmentations_with_baseimageaugmentationlayer.py

```diff
@@ -0,0 +1,435 @@
+
+layer = RandomBlueTint()
+augmented = layer(elephants)
+imshow(augmented.numpy())
+
+"""Looks great!  We can also call our layer on batched inputs:"""
+
+layer = RandomBlueTint()
+augmented = layer(tf.expand_dims(elephants, axis=0))
+imshow(augmented.numpy()[0])
+
+"""## Adding Random Behavior with the `FactorSampler` API.
+
+Most of the time, an image augmentation technique should not do the same thing on every
+invocation of the layer's `__call__` method.
+KerasCV offers the `FactorSampler` API to allow users to provide configurable random
+distributions.
+"""
+
+
+class RandomBlueTint(keras_cv.layers.BaseImageAugmentationLayer):
+    """RandomBlueTint randomly applies a blue tint to images.
+
+    Args:
+      factor: A tuple of two floats, a single float or a
+        `keras_cv.FactorSampler`. `factor` controls the extent to which the
+        image is blue shifted. `factor=0.0` makes this layer perform a no-op
+        operation, while a value of 1.0 uses the degenerated result entirely.
+        Values between 0 and 1 result in linear interpolation between the original
+        image and a fully blue image.
+        Values should be between `0.0` and `1.0`.  If a tuple is used, a `factor` is
+        sampled between the two values for every image augmented.  If a single float
+        is used, a value between `0.0` and the passed float is sampled.  In order to
+        ensure the value is always the same, please pass a tuple with two identical
+        floats: `(0.5, 0.5)`.
+    """
+
+    def __init__(self, factor, **kwargs):
+        super().__init__(**kwargs)
+        self.factor = preprocessing_utils.parse_factor(factor)
+
+    def augment_image(self, image, transformation=None):
+        [*others, blue] = tf.unstack(image, axis=-1)
+        blue_shift = self.factor() * 255
+        blue = tf.clip_by_value(blue + blue_shift, 0.0, 255.0)
+        return tf.stack([*others, blue], axis=-1)
+
+
+"""Now, we can configure the random behavior of ou `RandomBlueTint` layer.
+We can give it a range of values to sample from:
+
+"""
+
+many_elephants = tf.repeat(tf.expand_dims(elephants, axis=0), 9, axis=0)
+layer = RandomBlueTint(factor=0.5)
+augmented = layer(many_elephants)
+gallery_show(augmented.numpy())
+
+"""Each image is augmented differently with a random factor sampled from the range
+`(0, 0.5)`.
+
+We can also configure the layer to draw from a normal distribution:
+"""
+
+many_elephants = tf.repeat(tf.expand_dims(elephants, axis=0), 9, axis=0)
+factor = keras_cv.NormalFactorSampler(
+    mean=0.3, standard_deviation=0.1, min_value=0.0, max_value=1.0
+)
+layer = RandomBlueTint(factor=factor)
+augmented = layer(many_elephants)
+gallery_show(augmented.numpy())
+
+"""As you can see, the augmentations now are drawn from a normal distributions.
+There are various types of `FactorSamplers` including `UniformFactorSampler`,
+`NormalFactorSampler`, and `ConstantFactorSampler`.  You can also implement you own.
```

Comment:
This level of granularity will definitely help researchers to do experiments around augmentation with a lot more flexibility.